File: Go/098-validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func inorderTraversal(root *TreeNode) []int {
    var stack []*TreeNode
    var res []int
    for root != nil || len(stack) != 0 {
        if root != nil {
            stack, root = append(stack, root), root.Left
        } else {
            root, stack = stack[len(stack)- 1], stack[:len(stack) - 1]
            res, root = append(res, root.Val), root.Right
        }
    }
    return res
}

func isValidBST(root *TreeNode) bool {
    list := inorderTraversal(root)
    for i := 1; i < len(list); i++ {
        if list[i] <= list[i - 1] {
            return false
        }
    }
    return true
}

File: Python/098-validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        """
        这里不能简单地使用递归, 因为要保证节点的左子树只包含小于当前节点的数
        要保证节点的右子树只包含大于当前节点的数
        """
        def inorderTraversal(root: TreeNode) -> List[int]:
            res = []
            stack = []
            while root or stack:
                #将所有的子节点入栈
                if root:
                    stack.append(root)
                    root = root.left
                else:
                    #按根-右的顺序访问
                    root = stack.pop()
                    res.append(root.val)
                    root = root.right
            return res
        TreeList = inorderTraversal(root)
        for i in range(1, len(TreeList)):
            if TreeList[i] <= TreeList[i - 1]:
                return False
        return True